design/authenticators/application-identity/application-identity.md

```diff
@@ -0,0 +1,464 @@
+ about authenticating hosts is relevant also for authenticating users
+ . Whenever I say one of `host`, `user` or `role` it has the same meaning.
+  
+## Introduction
+
+In several authenticators we require hosts to contain some data which define
+their identity. This data is referred to as Application Identity. 
+The Application Identity gives the authenticator another layer of
+authentication, where it can authenticate applications using not only their
+Conjur identity and permissions (e.g host can `authenticate` with the
+authenticator `webservice`) but also their native identity (e.g Azure
+resource characteristics)
+ 
+ For example, in the Azure Authenticator the Application Identity
+  includes:
+   - Subscription ID
+   - Resource Group
+   - System-Assigned Identity (optional)
+   - User-Assigned Identity (optional)
+ 
+A host that will authenticate using the Azure authenticator will be defined
+ as follows:
+ ```
+- !host 
+  id: test-app 
+  annotations: 
+    authn-azure/subscription-id: test-subscription 
+    authn-azure/resource-group: test-group 
+``` 
+
+In the example above, the Application Identity is stored as annotations.
+In the Kubernetes Authenticator the Application Identity can be stored either
+in the host's ID or in its annotations. 
+For example:
+ ```
+- !host
+  id: test-app-service-account
+  annotations:
+    authn-k8s/namespace: some-namespace
+    authn-k8s/service-account: some-service-account
+
+- !host
+  id: some-namespace/service_account/some-service-account
+```
+
+In the example above, both hosts have the same Application Identity. One
+ stores it in its annotation and the other in its ID.
+
+## Issue Description
+
+When we add a new authenticator to Conjur, we need to add the mechanism
+of validating the host's Application Identity. Currently we have no generic
+way to do it and because of that it is less straight-forward to add a new
+authenticator. An example of what a generic implementation is can be seen in
+the [Authenticator Status Check](../authenticators-status). As you can see
+in the [Implementing Authenticator Status API
+](../authenticators-status/authn-status-new-implem.md) doc, it is very simple
+to add a new Status Check to an Authenticator. This happens because the flow
+starts with general checks that are relevant to all authenticators, and all
+that needs to be done in order to add a new Status Check is to implement a
+method named `status` in the relevant `Authenticator` class, and create a
+ Command Class that will perform the actual validations. 
+ 
+However, when validating the Application Identity we currently have no
+generic mechanism, as can be seen in the implementations for the [Azure
+Authenticator](/app/domain/authentication/authn_azure/validate_application_identity.rb) and 
+[Kubernetes Authenticator](/app/domain/authentication/authn_k8s/validate_application_identity.rb), where a lot of code is duplicated with
+ minor changes. 
```

Comment:
the duplicated code is (showing snippets from azure but they appear also in k8s):
```
def initialize(role_annotations:, service_id:, logger:)
        @role_annotations = role_annotations
        @service_id       = service_id
        @logger           = logger
      end

      def constraints
        @constraints ||= {
          subscription_id:          constraint_value("subscription-id"),
          resource_group:           constraint_value("resource-group"),
          user_assigned_identity:   constraint_value("user-assigned-identity"),
          system_assigned_identity: constraint_value("system-assigned-identity")
        }.compact
      end

# check the `service-id` specific constraint first to be more granular
      def constraint_value constraint_name
        annotation_value("authn-azure/#{@service_id}/#{constraint_name}") ||
          annotation_value("authn-azure/#{constraint_name}")
      end

      def annotation_value name
        annotation = @role_annotations.find { |a| a.values[:name] == name }

        # return the value of the annotation if it exists, nil otherwise
        if annotation
          @logger.debug(LogMessages::Authentication::RetrievedAnnotationValue.new(name))
          annotation[:value]
        end
      end

# validating that the annotations listed for the Conjur resource align with the permitted Azure constraints
      def validate_azure_annotations_are_permitted
        validate_prefixed_permitted_annotations("authn-azure/")
        validate_prefixed_permitted_annotations("authn-azure/#{@service_id}/")
      end

      # check if annotations with the given prefix is part of the permitted list
      def validate_prefixed_permitted_annotations prefix
        @logger.debug(LogMessages::Authentication::ValidatingAnnotationsWithPrefix.new(prefix))

        prefixed_annotations(prefix).each do |annotation|
          annotation_name = annotation[:name]
          next if prefixed_permitted_constraints(prefix).include?(annotation_name)
          raise Err::ConstraintNotSupported.new(annotation_name.gsub(prefix, ""), permitted_constraints)
        end
      end

      def prefixed_annotations prefix
        @role_annotations.select do |a|
          annotation_name = a.values[:name]

          annotation_name.start_with?(prefix) &&
            # verify we take only annotations from the same level
            annotation_name.split('/').length == prefix.split('/').length + 1
        end
      end

      # add prefix to all permitted constraints
      def prefixed_permitted_constraints prefix
        permitted_constraints.map { |k| "#{prefix}#{k}" }
      end

      def permitted_constraints
        @permitted_constraints ||= %w(
          subscription-id resource-group user-assigned-identity system-assigned-identity
        )
      end

 def validate_constraint_exists constraint
        unless application_identity.constraints[constraint]
          raise Err::RoleMissingConstraint, annotation_type_constraint(constraint)
        end
      end

 # validates that the application identity doesn't include logical constraint
      # combinations (e.g user_assigned_identity & system_assigned_identity)
      def validate_constraint_combinations
        identifiers = %i(user_assigned_identity system_assigned_identity)

        identifiers_constraints = application_identity.constraints.keys & identifiers
        unless identifiers_constraints.length <= 1
          raise Errors::Authentication::IllegalConstraintCombinations, annotation_type_constraints(identifiers_constraints)
        end
      end

def validate_token_identity_matches_annotations
        application_identity.constraints.each do |constraint|
          annotation_type  = constraint[0].to_s
          annotation_value = constraint[1]
          unless annotation_value == @token_identity[annotation_type.to_sym]
            raise Err::InvalidApplicationIdentity, annotation_type_constraint(annotation_type)
          end
        end
        @logger.debug(Log::ValidatedApplicationIdentity.new)
      end

      def annotation_type_constraints constraints
        constraints.map { |constraint| annotation_type_constraint(constraint) }
      end

      # converts the constraint to be in annotation style (e.g resource-group
      # instead of resource_group) to enhance supportability
      def annotation_type_constraint constraint
        constraint.to_s.tr('_', '-')
      end
```

This is pretty much the whole code of authn-azure's application identity validation and it is implemented also in authn-k8s